DBPW - Update docs with password policies & new Database interface (#10138)
hashicorp/vault: website/pages/docs/secrets/databases/custom.mdx

	// established by the backend.
	Close() error
}
```

Each of the request and response objects can also be found in `sdk/database/dbplugin/v5/database.go`.

In each of the requests, you will see a `Statements` object. This object represents the set of
commands to run for that particular operation. For the `NewUser` function, this is a set of
commands to create the user (and often set permissions for that user). These statements are
from the following fields in the API:

| API Argument               | Request Object                                     |
| -------------------------- | -------------------------------------------------- |
| `creation_statements`      | `NewUserRequest.Statements.Commands`               |
| `revocation_statements`    | `DeleteUserRequest.Statements.Commands`            |
| `rollback_statements`      | `NewUserRequest.RollbackStatements.Commands`       |
| `renew_statements`         | `UpdateUserRequest.Expiration.Statements.Commands` |
| `rotation_statements`      | `UpdateUserRequest.Password.Statements.Commands`   |
| `root_rotation_statements` | `UpdateUserRequest.Password.Statements.Commands`   |

> In many of the built in plugins, they replace `{{name}}` (or `{{username`), `{{password}}`,

Reviewer: ```suggestion
In many of the built in plugins, they replace `{{name}}` (or `{{username}}`), `{{password}}`,
```

Author: 👍

user specified as the configuration for the plugin. Your plugin should use this
data to make connections to the database. The response object contains a similar configuration
map. The response object should contain the configuration map that should be saved within Vault.
This allows the plugin to manipulate the configuration prior to saving it.

It is also passed a boolean value indicating if your plugin should initialize a connection to
the database during the `Initialize` call. This function is called when the configuration is
written. This allows the user to know whether the configuration is valid and able to connect
to the database in question. If this is set to false, no connection should be made during the
`Initialize` call, but subsequent calls to the other functions will need to open a connection.

## Serving your plugin

The plugin runs as a separate binary outside of Vault, so the plugin itself will need a `main`
function. Use the `Serve` function within `sdk/database/dbplugin/v5` to serve your plugin. You
will also need to pass some TLS configuration information that Vault uses when initializing the
plugin. Below is an example setup:

```go
package main

import (

> 	"github.com/hashicorp/vault/api/plugins"

Reviewer: Not sure if intentional, but spacing is set to 1 space on per tab in here.

Author: <img src=https://media.giphy.com/media/qxtxlL4sFFle/giphy.gif height=150>

That's a tab in every other view except github's diff view 0.o

| `renew_statements`         | `UpdateUserRequest.Expiration.Statements.Commands` |
| `rotation_statements`      | `UpdateUserRequest.Password.Statements.Commands`   |
| `root_rotation_statements` | `UpdateUserRequest.Password.Statements.Commands`   |

In many of the built in plugins, they replace `{{name}}` (or `{{username}}`), `{{password}}`,
and/or `{{expiration}}` with the associated values. It is up to your plugin to perform these
string replacements. There is a helper function located in `sdk/database/helper/dbutil`
called `QueryHelper` that assists in doing this string replacement. You are not required to
use it, but it will make your plugin's behavior consistent with the built in plugins.

The `InitializeRequest` object contains a map of keys to values. This data is what the
user specified as the configuration for the plugin. Your plugin should use this
data to make connections to the database. The response object contains a similar configuration
map. The response object should contain the configuration map that should be saved within Vault.
This allows the plugin to manipulate the configuration prior to saving it.

> It is also passed a boolean value indicating if your plugin should initialize a connection to

Reviewer: I think we should rephrase this sentence a bit and directly mention `IntializeRequest.VerifyConnection` as the boolean value that controls this behavior.

Author: :+1:

Database plugin.

## Plugin Interface

All plugins for the database secrets engine must implement the same interface. This interface
is found in `sdk/database/dbplugin/v5/database.go`

```go
type Database interface {
	// Initialize the database plugin. This is the equivalent of a constructor for the
	// database object itself.
	Initialize(ctx context.Context, req InitializeRequest) (InitializeResponse, error)

	// NewUser creates a new user within the database. This user is temporary in that it
	// will exist until the TTL expires.
	NewUser(ctx context.Context, req NewUserRequest) (NewUserResponse, error)

	// UpdateUser updates an existing user within the database.
	UpdateUser(ctx context.Context, req UpdateUserRequest) (UpdateUserResponse, error)

	// DeleteUser from the database. This should not error if the user didn't
	// exist prior to this call.
	DeleteUser(ctx context.Context, req DeleteUserRequest) (DeleteUserResponse, error)

	// Type returns the Name for the particular database backend implementation.
	// This type name is usually set as a constant within the database backend
	// implementation, e.g. "mysql" for the MySQL database backend. This is used

> 	// for things like metrics and logging. No behavior is switched on this.

Reviewer: ```suggestion
	// for things like metrics and logging.
```

I think the last bit is applicable now, but hard to guarantee/enforce.

Author: I copy pasted this from the current code to keep consistency. This function has never had any behavior based on it as it is just used as an identifier.

Reviewer: Fair enough. I guess we do control when and where we call this, so as long as we are mindful of its usage then this may remain to be the case.

	Type() (string, error)

	// Close attempts to close the underlying database connection that was
	// established by the backend.
	Close() error
}
```

> Each of the request and response objects can also be found in `sdk/database/dbplugin/v5/database.go`.

Reviewer 2: Is there existing precedent for representing paths to code this way? 2 considerations:

1. Unrooted paths are sometimes confusing
2. This path makes more sense in the context of reading docs in the repo than the website. For someone reading the docs on the website, this path may be doubly confusing because it's both unrooted and it is in no way relative to the website.

OTOH I can see why you've not used absolute paths, because that would require using a branch or blob name, which would either link rot or quickly become out of date. Overall I think what you've done is probably my preferred option, but is there anything we can do to help clarify? Previously I've seen people use e.g. `vault.git/sdk/...`, which has its own flaws but another option.

Author: I don't think there is precedence for this.

1. Not sure what we can do for this. This page is explicitly for people who want to write a custom plugin and need to have knowledge about writing Go.
2. I disagree. The path is to a `.go` file and is in reference to writing a new plugin. I feel the context makes that clearer.

Yea, I agree there are alternatives to this but none of them are particularly good. Since the `sdk` package isn't going anywhere and we try to remain backwards compatible in that package, I feel pretty comfortable with this.

> In each of the requests, you will see a `Statements` object. This object represents the set of

Reviewer 2: ```suggestion
In each of the requests, you will see at least 1 `Statements` object. This object represents the set of
```

nit: `UpdateUserRequest` has multiple, and the fact they weren't in the top level was slightly confusing for me initially.

Author: Good call. I also added a note here that `UpdateUserRequest` has them in sub-fields.

commands to run for that particular operation. For the `NewUser` function, this is a set of
commands to create the user (and often set permissions for that user). These statements are
from the following fields in the API:

| API Argument               | Request Object                                     |
| -------------------------- | -------------------------------------------------- |
| `creation_statements`      | `NewUserRequest.Statements.Commands`               |
| `revocation_statements`    | `DeleteUserRequest.Statements.Commands`            |
| `rollback_statements`      | `NewUserRequest.RollbackStatements.Commands`       |
| `renew_statements`         | `UpdateUserRequest.Expiration.Statements.Commands` |
| `rotation_statements`      | `UpdateUserRequest.Password.Statements.Commands`   |

> | `root_rotation_statements` | `UpdateUserRequest.Password.Statements.Commands`   |

Reviewer 2: Do we state anywhere what happens if you set both `rotation_statements` and `root_rotation_statements`?

Separately to this PR, perhaps we should mark `root_rotation_statements` as deprecated?

Author: `root_rotation_statements` is not deprecated. The DB engine uses it when the `/database/rotate-root/:name` endpoint is called, whereas `rotation_statements` is used during a static credential rotation. The DB engine selects which one based on what action it is performing.

Reviewer 2: Ah I read too much into how the config was actually stored from this table. Makes sense, thanks for clearing that up for me.

2. An easy way to see what functions that need to be implemented is to put the following as a
   global variable within your package: `var _ dbplugin.Database = (*MyDatabase)(nil)`. This
   will fail to compile if the `MyDatabase` type does not adhere to the `dbplugin.Database` interface.
3. Update the initialization function. This step can be a little confusing since there are
   three functions to keep in mind: In the version 4 interface, there was `Initialize` and
   `Init`. The `Initialize` function was replaced with the `Init` function and was not used
   some time before Vault 1.6. It was kept in the interface originally for backwards
   compatibility. The version 5 interface has an `Initialize` function that will need to replace
   the `Init` function. The two arguments passed into `Init` (`config map[string]interface{}`
   and `verifyConnection bool`) are included in the `InitializeRequest` object and should be
   referenced accordingly. The verison 4 `Init` function returned a `map[string]interface{}`
   indicating the configuration that should be saved in Vault. This is wrapped inside the
   `InitializeResponse` object within the version 5 interface.

> 4. Update `CreateUser` to `NewUser`. The `NewUserRequest` object contains the username and

Reviewer: Might be worth explicitly calling out this change in behavior from v4 to v5. As in, v4 used to leave password generation up to the plugin which then gets returned back to Vault, but v5 now accepts one from Vault/the request object which then consumes it.

Author: Good call. I'll put it as a notification block above the steps as it applies to both `NewUser` and `UpdateUser`.

   `CreateUser` generated a password and returned it, `SetCredentials` receives a password
   via a configuration struct and returns it, and `RotateRootCredentials` generated a
   password and was expected to return an updated copy of its entire configuration
   with the new password.
4. The `SetCredentials` and `RotateRootCredentials` used for static credential rotation,
   and rotating the root user's credentials respectively were essentially the same operation:
   change a user's password. The only practical difference was which user it was referring
   to. This was especially evident when `SetCredentials` was used when rotating root
   credentials (unless static credential rotation wasn't supported by the plugin in question).

The new interface is roughly modeled after a [gRPC](https://grpc.io/) interface. It has improved
future compatibility by not requiring changes to the interface definition to add additional data
in the requests or responses. It also simplifies the interface by merging several into a single
function call.

### Upgrading your custom database

Vault 1.6 supports both version 4 and version 5 database plugins. The support for version 4

> plugins will be removed in a future release. If you are using a version 4 custom database

Reviewer 2: Although it's implied in the instructions below, I think it might be helpful for users if we explicitly tell them how to determine whether their plugin is version 4 or 5 (i.e. check the `sdk` import paths used in its source, A or B)

Reviewer 2: Or alternatively, augment some of the references to the fairly abstract "version 4" and "version 5" with concrete references to the interfaces implemented.

Author: Good call. I've added a small section describing some ways that a user can tell if they are on v4 or v5 as well as references to the files where the interface definitions are.

plugin, the following are basic instructions for upgrading to version 5.

-> In version 4, password generation was the responsibility of the plugin via the
   `GenerateCredentials` function. This is no longer the case with version 5. Vault is
   responsible for generating passwords and passing them to the plugin via
   `NewUserRequest.Password` and `UpdateUserRequest.Password.NewPassword`.

1. Change the import path from `github.com/hashicorp/vault/sdk/database/dbplugin` to
   `github.com/hashicorp/vault/sdk/database/dbplugin/v5`. The package name is the same, so any
   references to `dbplugin` can remain as long as those symbols exist within the new package
   (such as the `Serve` function).

> 2. An easy way to see what functions that need to be implemented is to put the following as a

Author: :+1:

   global variable within your package: `var _ dbplugin.Database = (*MyDatabase)(nil)`. This
   will fail to compile if the `MyDatabase` type does not adhere to the `dbplugin.Database` interface.
3. Update the initialization function. This step can be a little confusing since there are
   three functions to keep in mind: In the version 4 interface, there was `Initialize` and
   `Init`. The `Initialize` function was replaced with the `Init` function and was not used
   some time before Vault 1.6. It was kept in the interface originally for backwards
   compatibility. The version 5 interface has an `Initialize` function that will need to replace

>    the `Init` function. The two arguments passed into `Init` (`config map[string]interface{}`

Reviewer 2: This block is quite hard to follow. I think some of the historical context could be pulled out into the background section above to leave this bullet as just the imperative instructions required to migrate.

Author: Updated. Let me know how it reads.

Reviewer 2: Way clearer now 💯  nice job
